presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java

```diff
@@ -140,6 +141,21 @@ public void testGenericLiteral()
         assertGenericLiteral("foo");
     }
 
+    @Test
+    public void testBinaryLiteral()
+        throws Exception
+    {
+        assertExpression("x' '", new BinaryLiteral(""));
+        assertExpression("x''", new BinaryLiteral(""));
+
+        // forms such as "X 'a b' " may look like BinaryLiteral
+        // but they do not pass the syntax rule for BinaryLiteral
+        // but instead conform to TypeConstructor, which generates a GenericLiteral expression
+        assertExpression("X 'a b'", new GenericLiteral("X", "a b"));
+        assertExpression("X'a b c'", new GenericLiteral("X", "a b c"));
```

Comment:
The error message the user gets when he/she accidentally puts an odd number of characters into BinaryLiteral will appear cryptic to the user. Let's change the parser to accept `X'<any thing>'`. Put a comment there that explains why we have to do that. And then do the more complex grammar check in the constructor of `BinaryLiteral`.